File: antlropt/org/perses/antlr/pnf/PlusIntroducerLeftPass.java
package org.perses.antlr.pnf;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.collect.HashMultimap;
import com.google.common.collect.ImmutableCollection;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Iterables;
import org.perses.antlr.RuleType;
import org.perses.antlr.ast.AbstractPersesRuleElement;
import org.perses.antlr.ast.AstEdit;
import org.perses.antlr.ast.AstTag;
import org.perses.antlr.ast.PersesPlusAst;
import org.perses.antlr.ast.PersesRuleReferenceAst;
import org.perses.antlr.ast.PersesSequenceAst;
import org.perses.antlr.ast.PersesStarAst;
import org.perses.antlr.ast.RuleNameRegistry.RuleNameHandle;

import java.util.ArrayList;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

import static com.google.common.base.Preconditions.checkState;

public class PlusIntroducerLeftPass extends AbstractPnfPass {
  @Override
  public ImmutableRuleDefMap process(ImmutableRuleDefMap grammar) {
    MutableRuleDefMap mutable = grammar.createMutable();
    PlusIntroducerEdit edit = new PlusIntroducerEdit(grammar);
    ArrayList<RuleEditTriple> toUpdate = new ArrayList<>();
    for (Map.Entry<RuleNameHandle, AbstractPersesRuleElement> entry : mutable.entries()) {
      final RuleNameHandle name = entry.getKey();
      final AbstractPersesRuleElement oldDef = entry.getValue();
      checkState(oldDef.getTag() != AstTag.ALTERNATIVE_BLOCK);
      Optional<AbstractPersesRuleElement> newDef = edit.apply(oldDef);
      if (!newDef.isPresent()) {
        continue;
      }
      toUpdate.add(new RuleEditTriple(name, oldDef, newDef.get()));
    }
    toUpdate.forEach(triple -> triple.applyTo(mutable));
    edit.toAdd.forEach(mutable::put);
    return grammar.createWithParserDefs(mutable);
  }

  @VisibleForTesting
  static class PlusIntroducerEdit extends AstEdit {

    private final ImmutableRuleDefMap defMap;

    @VisibleForTesting
    final HashMultimap<RuleNameHandle, AbstractPersesRuleElement> toAdd = HashMultimap.create();

    PlusIntroducerEdit(ImmutableRuleDefMap defMap) {
      this.defMap = defMap;
    }

    @Override
    protected Optional<AbstractPersesRuleElement> internalApply(
        AbstractPersesRuleElement element, final boolean isRoot) {
      AstTag tag = element.getTag();
      if (tag != AstTag.SEQUENCE) {
        return Optional.empty();
      }
      final int childCount = element.getChildCount();
      checkState(childCount > 1);
      PersesSequenceAst sequence = (PersesSequenceAst) element;
      final ArrayList<AbstractPersesRuleElement> childrenCopy =
          new ArrayList<>(sequence.getChildren());
      for (int i = 0; i < childCount; ++i) {
        final AbstractPersesRuleElement child = sequence.getChild(i);
        final AstTag childTag = child.getTag();
        checkState(
            !childTag.isQuantifier(),
            "Quantified node is not expected here. "
                + "You can run a normalization pass before this pass to fix the problem.");
        if (childTag != AstTag.RULE_REF) {
          continue;
        }
        final RuleNameHandle ruleName = ((PersesRuleReferenceAst) child).getRuleNameHandle();
        Optional<PersesStarAst> starAst = getStarIfIsKleeneStarRule(ruleName);
        if (!starAst.isPresent()) {
          continue;
        }
        AbstractPersesRuleElement quantified = starAst.get().getBody();
        i = transformStarToPlus(sequence, i, childrenCopy, quantified, ruleName, starAst.get());
      }
      ImmutableList<AbstractPersesRuleElement> newChildren =
          childrenCopy.stream().filter(Objects::nonNull).collect(ImmutableList.toImmutableList());
      checkState(element.getTag() == AstTag.SEQUENCE);
      checkState(newChildren.size() > 0);
      if (newChildren.size() == 1) {
        return Optional.of(Iterables.getOnlyElement(newChildren));
      } else {
        return Optional.of(new PersesSequenceAst(newChildren));
      }
    }

    @VisibleForTesting
    int transformStarToPlus(
        final PersesSequenceAst element,
        int currentIndexOfStar,
        ArrayList<AbstractPersesRuleElement> newChildren,
        AbstractPersesRuleElement bodyToMatch,
        RuleNameHandle starRuleName,
        PersesStarAst starAst) {
      final ImmutableList<AbstractPersesRuleElement> bodySequence =
          bodyToMatch.getTag() == AstTag.SEQUENCE
              ? ((PersesSequenceAst) bodyToMatch).getChildren()
              : ImmutableList.of(bodyToMatch);

      if (reverseMatch(element, currentIndexOfStar - 1, bodySequence)) {
        newChildren.set(currentIndexOfStar, mergeStarIntoPlus(starRuleName, starAst));
        for (int i = 0, size = bodySequence.size(); i < size; ++i) {
          newChildren.set(currentIndexOfStar - i - 1, null);
        }
        return currentIndexOfStar;
      }
      if (match(element, currentIndexOfStar + 1, bodySequence)) {
        newChildren.set(currentIndexOfStar, mergeStarIntoPlus(starRuleName, starAst));
        for (int i = 0, size = bodySequence.size(); i < size; ++i) {
          newChildren.set(currentIndexOfStar + i + 1, null);
        }
        return currentIndexOfStar + bodySequence.size();
      }
      if (bodyToMatch.getTag() == AstTag.RULE_REF) {
        ImmutableList<AbstractPersesRuleElement> defs =
            defMap.getRuleDefinition(((PersesRuleReferenceAst) bodyToMatch).getRuleNameHandle());
        if (defs.size() != 1) {
          return currentIndexOfStar;
        }
        AbstractPersesRuleElement def = Iterables.getOnlyElement(defs);
        return transformStarToPlus(
            element, currentIndexOfStar, newChildren, def, starRuleName, starAst);
      }
      return currentIndexOfStar;
    }

    @VisibleForTesting
    static boolean reverseMatch(
        PersesSequenceAst sequence,
        int startIndex,
        ImmutableList<AbstractPersesRuleElement> toMatchList) {
      final int length = toMatchList.size();
      return match(sequence, startIndex - length + 1, toMatchList);
    }

    @VisibleForTesting
    static boolean match(
        PersesSequenceAst sequence,
        int startIndex,
        ImmutableList<AbstractPersesRuleElement> toMatchList) {
      if (startIndex < 0 | startIndex >= sequence.getChildCount()) {
        return false;
      }
      for (int i = 0, size = toMatchList.size(); i < size; ++i) {
        final AbstractPersesRuleElement toMatch = toMatchList.get(i);
        final int indexInSeq = startIndex + i;
        if (indexInSeq >= sequence.getChildCount()) {
          return false;
        }
        if (!toMatch.isEquivalent(sequence.getChild(indexInSeq))) {
          return false;
        }
      }
      return true;
    }

    private PersesRuleReferenceAst mergeStarIntoPlus(
        RuleNameHandle starRuleName, PersesStarAst starAst) {
      final RuleNameHandle plusRuleName =
          starRuleName.createAuxiliaryRuleName(RuleType.KLEENE_PLUS);
      //      toRemove.put(starRuleName, starAst);
      toAdd.put(
          plusRuleName,
          starAst.isGreedy()
              ? PersesPlusAst.createGreedy(starAst.getBody())
              : PersesPlusAst.createNonGreedy(starAst.getBody()));
      return new PersesRuleReferenceAst(plusRuleName);
    }

    @VisibleForTesting
    Optional<PersesStarAst> getStarIfIsKleeneStarRule(RuleNameHandle ruleName) {
      ImmutableList<AbstractPersesRuleElement> defs = defMap.getRuleDefinition(ruleName);
      if (defs.size() != 1) {
        return Optional.empty();
      }
      AbstractPersesRuleElement def = Iterables.getOnlyElement(defs);
      if (def.getTag() != AstTag.STAR) {
        return Optional.empty();
      }
      return Optional.of((PersesStarAst) Iterables.getOnlyElement(defs));
    }
  }
}

File: antlropt/src/org/perses/antlr/pnf/PlusIntroducerLeftPass.kt
/*
 * Copyright (C) 2018-2024 University of Waterloo.
 *
 * This file is part of Perses.
 *
 * Perses is free software; you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software
 * Foundation; either version 3, or (at your option) any later version.
 *
 * Perses is distributed in the hope that it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * Perses; see the file LICENSE.  If not see <http://www.gnu.org/licenses/>.
 */
package org.perses.antlr.pnf

import com.google.common.annotations.VisibleForTesting
import com.google.common.collect.ImmutableList
import com.google.common.collect.LinkedHashMultimap
import org.perses.antlr.RuleType
import org.perses.antlr.ast.AbstractPersesRuleElement
import org.perses.antlr.ast.AstEdit
import org.perses.antlr.ast.AstTag
import org.perses.antlr.ast.PersesAlternativeBlockAst
import org.perses.antlr.ast.PersesGrammar
import org.perses.antlr.ast.PersesPlusAst.Companion.createGreedy
import org.perses.antlr.ast.PersesPlusAst.Companion.createNonGreedy
import org.perses.antlr.ast.PersesRuleReferenceAst
import org.perses.antlr.ast.PersesRuleReferenceAst.Companion.create
import org.perses.antlr.ast.PersesSequenceAst
import org.perses.antlr.ast.PersesStarAst
import org.perses.antlr.ast.RuleNameRegistry.RuleNameHandle
import org.perses.antlr.ast.TransformDecision
import org.perses.util.exhaustive
import org.perses.util.toImmutableList

class PlusIntroducerLeftPass : AbstractPnfPass() {
  override fun processGrammar(
    grammar: GrammarPair,
  ): GrammarPair {
    val parserGrammar = grammar.parserGrammar ?: return grammar
    val mutable = MutableGrammar.createParserRulesFrom(parserGrammar)
    val edit = PlusIntroducerEdit(parserGrammar)
    val toUpdate = ArrayList<RuleEditTriple>()
    mutable.ruleNameAltPairSequence().forEach { (name, oldDef) ->
      check(oldDef.tag !== AstTag.ALTERNATIVE_BLOCK)
      val decision = edit.apply(oldDef)
      when (decision) {
        is TransformDecision.Keep -> return@forEach
        is TransformDecision.Replace ->
          toUpdate.add(RuleEditTriple(name, oldDef, decision.newValue))
        else -> TODO(decision.toString())
      }.exhaustive
    }
    toUpdate.forEach { it.applyTo(mutable) }
    edit.toAdd.forEach { key, value ->
      mutable.getAltBlock(key).addIfNotEquivalent(value)
    }
    return grammar.withNewParserGrammar(
      parserGrammar.copyWithNewParserRuleDefs(mutable.toParserRuleAstList()),
    )
  }

  @VisibleForTesting
  class PlusIntroducerEdit(private val grammar: PersesGrammar) : AstEdit() {
    @JvmField
    @VisibleForTesting
    val toAdd = LinkedHashMultimap
      .create<RuleNameHandle, AbstractPersesRuleElement>()

    public override fun internalApply(
      element: AbstractPersesRuleElement,
      isRoot: Boolean,
    ): TransformDecision.NonDeleteTransformDecision {
      val tag = element.tag
      if (tag !== AstTag.SEQUENCE) {
        return TransformDecision.Keep(element)
      }
      val childCount = element.childCount
      check(childCount > 1)
      val sequence = element as PersesSequenceAst
      val childrenCopy = ArrayList(sequence.children)
      var i = 0
      while (i < childCount) {
        val child = sequence.getChild(i)
        val childTag = child.tag
        check(childTag != null && !childTag.isQuantifier()) {
          "Quantified node is not expected here. " +
            "You can run a normalization pass before this pass to fix the problem."
        }
        if (childTag !== AstTag.RULE_REF) {
          ++i
          continue
        }
        val ruleName = (child as PersesRuleReferenceAst).ruleNameHandle
        val starAst = getStarIfIsKleeneStarRule(ruleName)
        if (starAst == null) {
          ++i
          continue
        }
        val quantified = starAst.body
        i = transformStarToPlus(sequence, i, childrenCopy, quantified, ruleName, starAst)
        ++i
      }
      val newChildren = childrenCopy
        .asSequence()
        .filter { it != null }
        .toImmutableList()
      check(element.tag === AstTag.SEQUENCE)
      check(newChildren.size > 0)
      val finalResult = if (newChildren.size == 1) {
        newChildren.single()
      } else {
        PersesSequenceAst(newChildren)
      }

      return if (finalResult.isEquivalent(sequence)) {
        // FIXME(cnsun): this is a bug, which needs to be fixed.
        //               check(!finalResult.isEquivalent(sequence)) {finalResult}
        TransformDecision.Keep(sequence)
      } else {
        TransformDecision.Replace(oldValue = sequence, newValue = finalResult)
      }
    }

    @VisibleForTesting
    fun transformStarToPlus(
      element: PersesSequenceAst,
      currentIndexOfStar: Int,
      newChildren: ArrayList<AbstractPersesRuleElement?>,
      bodyToMatch: AbstractPersesRuleElement,
      starRuleName: RuleNameHandle,
      starAst: PersesStarAst,
    ): Int {
      val bodySequence = if (bodyToMatch.tag === AstTag.SEQUENCE) {
        (bodyToMatch as PersesSequenceAst).children
      } else {
        ImmutableList.of(bodyToMatch)
      }
      if (reverseMatch(element, currentIndexOfStar - 1, bodySequence)) {
        newChildren[currentIndexOfStar] = mergeStarIntoPlus(starRuleName, starAst)
        var i = 0
        val size = bodySequence.size
        while (i < size) {
          newChildren[currentIndexOfStar - i - 1] = null
          ++i
        }
        return currentIndexOfStar
      }
      if (match(element, currentIndexOfStar + 1, bodySequence)) {
        newChildren[currentIndexOfStar] = mergeStarIntoPlus(starRuleName, starAst)
        var i = 0
        val size = bodySequence.size
        while (i < size) {
          newChildren[currentIndexOfStar + i + 1] = null
          ++i
        }
        return currentIndexOfStar + bodySequence.size
      }
      if (bodyToMatch.tag === AstTag.RULE_REF) {
        val ruleNameHandle = (bodyToMatch as PersesRuleReferenceAst).ruleNameHandle
        val ruleDef = grammar.getRuleDefinition(ruleNameHandle)!!
        if (ruleDef.body is PersesAlternativeBlockAst) {
          return currentIndexOfStar
        }
        return transformStarToPlus(
          element,
          currentIndexOfStar,
          newChildren,
          ruleDef.body,
          starRuleName,
          starAst,
        )
      }
      return currentIndexOfStar
    }

    private fun mergeStarIntoPlus(
      starRuleName: RuleNameHandle,
      starAst: PersesStarAst,
    ): PersesRuleReferenceAst {
      val plusRuleName = starRuleName.createAuxiliaryRuleName(RuleType.KLEENE_PLUS)
      //      toRemove.put(starRuleName, starAst);
      toAdd.put(
        plusRuleName,
        if (starAst.isGreedy) createGreedy(starAst.body) else createNonGreedy(starAst.body),
      )
      return create(plusRuleName)
    }

    @VisibleForTesting
    fun getStarIfIsKleeneStarRule(ruleName: RuleNameHandle): PersesStarAst? {
      val ruleDef = grammar.getRuleDefinition(ruleName) ?: return null
      val body = ruleDef.body
      return if (body is PersesStarAst) {
        body
      } else {
        null
      }
    }

    companion object {
      @JvmStatic
      @VisibleForTesting
      fun reverseMatch(
        sequence: PersesSequenceAst,
        startIndex: Int,
        toMatchList: ImmutableList<AbstractPersesRuleElement>,
      ): Boolean {
        val length = toMatchList.size
        return match(sequence, startIndex - length + 1, toMatchList)
      }

      @JvmStatic
      @VisibleForTesting
      fun match(
        sequence: PersesSequenceAst,
        startIndex: Int,
        toMatchList: ImmutableList<AbstractPersesRuleElement>,
      ): Boolean {
        if (startIndex < 0 || startIndex >= sequence.childCount) {
          return false
        }
        var i = 0
        val size = toMatchList.size
        while (i < size) {
          val toMatch = toMatchList[i]
          val indexInSeq = startIndex + i
          if (indexInSeq >= sequence.childCount) {
            return false
          }
          if (!toMatch.isEquivalent(sequence.getChild(indexInSeq))) {
            return false
          }
          ++i
        }
        return true
      }
    }
  }
}
